toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -185,13 +313,206 @@ func resolveGraphNodes(dependencyGraph *pkggraph.PkgGraph, inputSummaryFile, out
+	var lookup *pkggraph.LookupNode
+
+	// Find the real build node in the graph we want to keep (we will be discarding the graph the node was passed in from so we can't use it)
+	if realBuildNode.Type != pkggraph.TypeBuild {
+		err = fmt.Errorf("node '%s' is not a build node, can't download delta RPM", realBuildNode)
+		return false, err
+	}
+
+	timestamp.StartEvent(fmt.Sprintf("downloading delta node %s", realBuildNode.VersionedPkg.Name), nil)
+	defer timestamp.StopEvent(nil)
+
+	lookup, err = realDependencyGraph.FindExactPkgNodeFromPkg(realBuildNode.VersionedPkg)
+	if err != nil {
+		err = fmt.Errorf("can't find build node '%s' in graph: %w", realBuildNode, err)
+		return false, err
+	}
+	if lookup == nil || lookup.RunNode == nil {
+		err = fmt.Errorf("can't find run lookup '%v' in graph", lookup)
+		return false, err
+	}
+
+	if lookup.BuildNode != realBuildNode {
+		err = fmt.Errorf("real build node '%v' does not match build node in the graph lookup '%v'", realBuildNode, lookup.BuildNode)
+		return false, err
+	}
+
+	realRunNode := lookup.RunNode
+	// We need to copy the node since the fetcher may on occasion choose a different package to provide something than
+	// we want. We only consider a delta node valid if the .rpm matches exactly. If we tryy and fail to download a delta
+	// the node would be left in a bad state.
+	nodeCopy := realBuildNode
+
+	// Get the final output path for the build node if we don't convert it to a delta node
+	originalRpmPath := realBuildNode.RpmPath
+	foundFinalRPM, _ := file.IsFile(originalRpmPath)
+
+	// Only download dependencies for delta RPMs if we don't already have the RPM in the out/RPMS folder
+	if !foundFinalRPM {
+		// Flag the node as a delta node now so the resolver knows this isn't a critical issue if it fails to resolve
+		nodeCopy.State = pkggraph.StateDelta
+		resolveErr := resolveSingleNode(cloner, nodeCopy, downloadDependencies, nil, make(map[string]bool), make(map[string]bool), deltaRpmDir)
+		// Failing to clone a dependency should not halt a build.
+		// The build should continue and attempt best effort to build as many packages as possible.
+		if resolveErr != nil {
+			logger.Log.Warnf("Can't find delta RPM to download for %s-%s: %s", nodeCopy.VersionedPkg.Name, nodeCopy.VersionedPkg.Version, resolveErr)
+			return false, nil
 		}
+		logger.Log.Tracef("Updating real node '%s' with info from newly cached delta node '%s'", realBuildNode, nodeCopy)
+
+		// Check that the RPM we are getting matches the expected out/RPM path using the base name of the file path
+		cachedRPMPath := nodeCopy.RpmPath
+		if filepath.Base(cachedRPMPath) != filepath.Base(originalRpmPath) {
+			logger.Log.Warnf("cached delta RPM '%s' does not match expected RPM '%s', skipping", filepath.Base(cachedRPMPath), filepath.Base(originalRpmPath))
+			return false, nil
+		} else {
+			logger.Log.Infof("Delta RPM found for '%s-%s'.", realBuildNode.VersionedPkg.Name, realBuildNode.VersionedPkg.Version)
+		}
```

Comment:
I think using `resolveSingleNode` may be an overkill here. From the `filepath.Base(cachedRPMPath) != filepath.Base(originalRpmPath)` check I understand, that we want to find the package, which matches in name with the one from the build node. If that's the case, then I think we can go straight to `cloner.Clone(downloadDependencies, rpmName)`, where `rpmName` is `originalRpmPath` without the `.rpm` suffix.

This way:
- We're sure to get the matching RPM name, assuming the package is available upstream, so the `filepath.Base(cachedRPMPath) != filepath.Base(originalRpmPath)` check is unnecessary, and
- we don't need to create and use a copy of `realBuildNode`.

Note: this change would require extracting part of the `cloner.Clone` function, so that we have a version working directly on strings, but we can talk about it and I'm happy to do that for you, if you'd like.

Reply:
Was able to get this working by adding a new `CloneRawPackageNames()` to rpmrepocloner.  Both that function and `Clone()` now utilize `cloneInternal()` which handles both cases.